nrf-softdevice/src/ble/advertisement_builder.rs

```diff
@@ -138,145 +160,163 @@ pub enum BasicService {
-        self.write(&[(N * S::SIZE) as u8 + 1, L::AD as u8]);
+    /// Add a list of 16-bit service uuids to the advertisement data.
+    pub const fn services_16(self, complete: ServiceList, services: &[BasicService]) -> Self {
+        let ad_type = match complete {
+            ServiceList::Incomplete => AdvertisementDataType::INCOMPLETE_16_SERVICE_LIST,
+            ServiceList::Complete => AdvertisementDataType::COMPLETE_16_SERVICE_LIST,
+        };
+
+        let mut res = self.write(&[(services.len() * 2) as u8 + 1, ad_type.to_u8()]);
+        let mut i = 0;
+        while i < services.len() {
+            res = res.write(&(services[i] as u16).to_le_bytes());
+            i += 1;
+        }
+        res
+    }
 
-        for service in services.list() {
-            service.render(&mut self);
+    /// Add a list of 128-bit service uuids to the advertisement data.
+    ///
+    /// Note that each UUID in the list needs to be in little-endian format, i.e. opposite to what you would
+    /// normally write UUIDs.
+    pub const fn services_128(self, complete: ServiceList, services: &[[u8; 16]]) -> Self {
+        let ad_type = match complete {
+            ServiceList::Incomplete => AdvertisementDataType::INCOMPLETE_128_SERVICE_LIST,
+            ServiceList::Complete => AdvertisementDataType::COMPLETE_128_SERVICE_LIST,
+        };
+
+        let mut res = self.write(&[(services.len() * 16) as u8 + 1, ad_type.to_u8()]);
+        let mut i = 0;
+        while i < services.len() {
+            res = res.write(&services[i]);
+            i += 1;
         }
+        res
+    }
 
-        self
+    /// Add a name to the advertisement data.
+    pub const fn short_name(self, name: &str) -> Self {
+        self.raw(AdvertisementDataType::SHORT_NAME, name.as_bytes())
     }
 
     /// Add a name to the advertisement data.
-    pub fn name<N: Name>(self, name: N) -> Self {
-        self.raw(N::AD, name.inner().as_bytes())
+    pub const fn full_name(self, name: &str) -> Self {
+        self.raw(AdvertisementDataType::FULL_NAME, name.as_bytes())
     }
 
     /// If the full name fits within the remaining space, it is used. Otherwise the short name is used.
     ///
     /// *Note: This modifier should be placed last.*
-    pub fn adapt_name(self, full: FullName, short: ShortName) -> Self {
-        let full_len = full.inner().len();
-
-        if self.ptr + full_len <= K {
-            self.name(full)
+    pub const fn adapt_name(self, name: &str) -> Self {
+        let p = self.ptr;
+        if p + 2 + name.len() <= K {
+            self.full_name(name)
         } else {
-            self.name(short)
+            let mut res = self.write(&[(K - p) as u8, AdvertisementDataType::SHORT_NAME.to_u8()]);
+            let mut i: usize = 0;
+            let bytes = name.as_bytes();
+            while res.ptr < K {
+                res.buf[res.ptr] = bytes[i];
+                res.ptr += 1;
+                i += 1;
+            }
+            res
```

Comment:
This must have made you want to pull your hair out. I *can't believe* how annoying this is in const land. This is 4 lines non-const, crazy. Good work!

My only suggestion is to update the docstring: "Adds the provided string as a name, truncating and typing as needed."